src/trace_queue.js

```diff
@@ -0,0 +1,196 @@
+        cancelTokenSource.cancel('Timeout sending batch!');
+    }, config.getConfig().sendTimeout);
+
+    return session.post(
+        config.getConfig().traceCollectorURL,
+        batchObject,
+        {
+            headers: { Authorization: `Bearer ${config.getConfig().token}` },
+            timeout: config.getConfig().sendTimeout,
+            cancelToken: cancelTokenSource.token,
+        }
+    ).then((res) => {
+        clearTimeout(handle);
+        utils.debugLog('[QUEUE] Batch posted!');
+        return res;
+    }).catch((err) => {
+        clearTimeout(handle);
+        if (err.config && err.config.data) {
+            utils.debugLog(`[QUEUE] Error sending trace. Batch size: ${err.config.data.length}`);
+        } else {
+            utils.debugLog(`[QUEUE] Error sending trace. Error: ${err}`);
+        }
+        utils.debugLog(`[QUEUE] ${err ? err.stack : err}`);
+        return err;
+    });
+}
+
+
+class TraceQueue extends EventEmitter {
+    constructor() {
+        super();
+        this.batchSender = postBatch;
+        this.traces = [];
+        this.updateConfig();
+        this.initQueue();
+    }
+
+    updateConfig() {
+        this.maxTraceWait = config.getConfig().maxTraceWait;
+        this.maxBatchSizeBytes = config.getConfig().maxBatchSizeBytes;
+        this.batchSize = config.getConfig().batchSize;
+    }
+
+    /**
+   * Init queue event listners
+   */
+    initQueue() {
+        this.removeAllListeners();
+        this.currentByteSize = 0;
+        this.flush();
+        this.on('traceQueued', () => {
+            if (this.byteSizeLimitReached()) {
+                utils.debugLog(`[QUEUE] Queue Byte size reached ${this.currentByteSize} Bytes, releasing batch...`);
+                this.releaseBatch(this.currentSize - 1);
+            } else if (this.batchSizeReached()) {
+                utils.debugLog(`[QUEUE] Queue size reached ${this.currentSize}, releasing batch... `);
+                this.releaseBatch();
+            }
+        });
+
+        this.on('batchReleased', (batch) => {
+            utils.debugLog(`[QUEUE] Queue Byte size reached ${this.currentByteSize} Bytes, releasing batch...`);
+            this.batchSender(batch);
+            this.emit('batchSent', batch);
+        });
+    }
+
+    /**
+   * Queue size getter
+   * @returns {Number}
+   */
+    get currentSize() {
+        return this.traces.length;
+    }
+
+
+    /**
+   * Checks if queue size reached batch size
+   * @returns {Boolean}
+   */
+    batchSizeReached() {
+        return this.currentSize === this.batchSize;
+    }
+
+    /**
+   * Checks if queue byte size reached its limit
+   * @returns {Boolean}
+   */
+    byteSizeLimitReached() {
+        return this.currentByteSize >= this.maxBatchSizeBytes;
+    }
+
+    /**
+   * add given trace byte size to total byte size
+   * @returns {Number}
+   */
+    addToCurrentByteSize(trace) {
+        this.currentByteSize += JSON.stringify(trace).length;
```

Comment:
`JSON. stringify` can be CPU expensive. Maybe we should save the traces as JSONs already?